data_structures/binary_tree/segment_tree_other.py

```diff
@@ -4,17 +4,29 @@
 function takes 2 values and returns a same type value
 """
 from collections.abc import Sequence
+from functools import reduce
 from queue import Queue
+from typing import TypeVar
+
+T = TypeVar("T", int, float)
 
 
 class SegmentTreeNode:
-    def __init__(self, start, end, val, left=None, right=None):
+    def __init__(
+        self,
+        start: int,
+        end: int,
+        val: T,
+        left: "SegmentTreeNode" = None,
+        right: "SegmentTreeNode" = None,
+    ):
```

Comment:
A function with no doctests?

Reply:
@cclauss The `_update_tree_range` is a private function, which isn't supposed to be used on its own. Instead, it is called by the `update_range` method, which is properly tested. So it looks like there is no need to test `_update_tree_range` specifically, but I will do it, if it's really necessary.

Reply:
Private functions can still generate garbage output.  When we want to know what is the root cause of a problem, it is helpful if all the components have tests.

Reply:
Okay, I will add tests for private methods as well.

Reply:
@cclauss Done.